moveit_ros/planning/trajectory_execution_manager/src/trajectory_execution_manager.cpp

```diff
@@ -236,63 +254,92 @@ void TrajectoryExecutionManager::receiveEvent(const std_msgs::StringConstPtr& ev
+                                      const ExecutionCompleteCallback& callback)
 {
   moveit_msgs::RobotTrajectory traj;
   traj.joint_trajectory = trajectory;
-  return push(traj, controllers);
+  return push(traj, controllers, callback);
 }
 
 bool TrajectoryExecutionManager::push(const moveit_msgs::RobotTrajectory& trajectory,
-                                      const std::vector<std::string>& controllers)
+                                      const std::vector<std::string>& controllers,
+                                      const ExecutionCompleteCallback& callback)
 {
-  if (!execution_complete_)
+  std::vector<moveit_msgs::RobotTrajectory> trajs = { trajectory };
+  return push(trajs, controllers, callback);
+}
+
+bool TrajectoryExecutionManager::push(const std::vector<moveit_msgs::RobotTrajectory>& trajectories,
+                                      const std::vector<std::string>& controllers,
+                                      const ExecutionCompleteCallback& callback)
+{
+  ROS_INFO_NAMED(LOGNAME, "Pushing new trajectory");
+  if (!execution_complete_ && !allow_continuous_execution_)
   {
     ROS_ERROR_NAMED(LOGNAME, "Cannot push a new trajectory while another is being executed");
     return false;
   }
 
-  TrajectoryExecutionContext* context = new TrajectoryExecutionContext();
-  if (configure(*context, trajectory, controllers))
+  std::vector<std::shared_ptr<TrajectoryExecutionContext>> contexts;
+
+  for (auto trajectory : trajectories)
   {
-    if (verbose_)
+    std::shared_ptr<TrajectoryExecutionContext> context = std::make_shared<TrajectoryExecutionContext>();
+    if (configure(*context, trajectory, controllers))
     {
-      std::stringstream ss;
-      ss << "Pushed trajectory for execution using controllers [ ";
-      for (const std::string& controller : context->controllers_)
-        ss << controller << " ";
-      ss << "]:" << std::endl;
-      for (const moveit_msgs::RobotTrajectory& trajectory_part : context->trajectory_parts_)
-        ss << trajectory_part << std::endl;
-      ROS_INFO_NAMED(LOGNAME, "%s", ss.str().c_str());
+      context->execution_complete_callback = callback;
+      if (verbose_)
+      {
+        std::stringstream ss;
+        ss << "Pushed trajectory for execution using controllers [ ";
+        for (const std::string& controller : context->controllers_)
+          ss << controller << " ";
+        ss << "]:" << std::endl;
+        for (const moveit_msgs::RobotTrajectory& trajectory_part : context->trajectory_parts_)
+          ss << trajectory_part << std::endl;
+        ROS_INFO_NAMED(LOGNAME, "%s", ss.str().c_str());
+      }
+      contexts.push_back(context);
+    }
+    else
+    {
+      context.reset();
+      last_execution_status_ = moveit_controller_manager::ExecutionStatus::ABORTED;
+      return false;
     }
-    trajectories_.push_back(context);
-    return true;
+  }
+
+  if (allow_continuous_execution_)
+  {
+    ROS_INFO_NAMED(LOGNAME, "Creating thread");
```

Comment:
I wonder if the validation should happen here already and basically the push being rejected (return false) if the trajectory cannot currently be executed?